File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/dto/ItemDto.java
package ai.ragu.charactermanager.dto;

import ai.ragu.charactermanager.enumeration.*;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.*;

import java.util.List;
import java.util.Map;


/**
 * An item
 */
@Data
@AllArgsConstructor
@NoArgsConstructor
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "code")
@ToString(onlyExplicitlyIncluded = true)
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class ItemDto implements Equipment {

    @JsonProperty("code")
    @JsonPropertyDescription("A unique identifier for this item")
    @ToString.Include
    @EqualsAndHashCode.Include
    private String code;

    @JsonProperty("name")
    @JsonPropertyDescription("The name for this item")
    private String name;

    @JsonProperty("description")
    @JsonPropertyDescription("The description for this item")
    private String description;

    @JsonProperty("traits")
    @JsonPropertyDescription("The traits associated to this item")
    private List<String> traits;

    @JsonProperty("usage")
    @JsonPropertyDescription("The usage of this item")
    private String usage;

    @JsonProperty("numActions")
    @JsonPropertyDescription("The number of actions required to use this item")
    private String numActions;

    @JsonProperty("types")
    @JsonPropertyDescription("The types of this item")
    private List<ItemTypeDto> types;

    @JsonProperty("activation")
    @JsonPropertyDescription("The activation modes of this item")
    private List<ActivationEnum> activation;

    @JsonProperty("bulk")
    @JsonPropertyDescription("The bulk of this item")
    private ItemBulkEnum bulk;

    @JsonProperty("price")
    @JsonPropertyDescription("The price of this item")
    private Map<CurrencyEnum, Integer> price;

    @JsonProperty("level")
    @JsonPropertyDescription("The level for this item")
    private int level;

    @JsonProperty("craftRequirements")
    @JsonPropertyDescription("The craft requirements of this item type")
    private String craftRequirements;
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/dto/ItemDto.kt
package ai.ragu.charactermanager.dto

import ai.ragu.charactermanager.enumeration.ActivationEnum
import ai.ragu.charactermanager.enumeration.CurrencyEnum
import ai.ragu.charactermanager.enumeration.ItemBulkEnum
import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyDescription
import com.fasterxml.jackson.annotation.ObjectIdGenerators.PropertyGenerator
import lombok.*

/**
 * An item
 */
@Data
@AllArgsConstructor
@NoArgsConstructor
@JsonIdentityInfo(generator = PropertyGenerator::class, property = "code")
@ToString(onlyExplicitlyIncluded = true)
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
class ItemDto : Equipment {
    @JsonProperty("code")
    @JsonPropertyDescription("A unique identifier for this item")
    @ToString.Include
    @EqualsAndHashCode.Include
    var code: String? = null

    @JsonProperty("name")
    @JsonPropertyDescription("The name for this item")
    var name: String? = null

    @JsonProperty("description")
    @JsonPropertyDescription("The description for this item")
    var description: String? = null

    @JsonProperty("traits")
    @JsonPropertyDescription("The traits associated to this item")
    var traits: List<String>? = null

    @JsonProperty("usage")
    @JsonPropertyDescription("The usage of this item")
    var usage: String? = null

    @JsonProperty("numActions")
    @JsonPropertyDescription("The number of actions required to use this item")
    var numActions: String? = null

    @JsonProperty("types")
    @JsonPropertyDescription("The types of this item")
    var types: List<ItemTypeDto>? = null

    @JsonProperty("activation")
    @JsonPropertyDescription("The activation modes of this item")
    var activation: List<ActivationEnum>? = null

    @JsonProperty("bulk")
    @JsonPropertyDescription("The bulk of this item")
    var bulk: ItemBulkEnum? = null

    @JsonProperty("price")
    @JsonPropertyDescription("The price of this item")
    var price: Map<CurrencyEnum, Int>? = null

    @JsonProperty("level")
    @JsonPropertyDescription("The level for this item")
    var level = 0

    @JsonProperty("craftRequirements")
    @JsonPropertyDescription("The craft requirements of this item type")
    var craftRequirements: String? = null
}